corehq/apps/accounting/tests/test_autopay_with_api.py

```diff
@@ -118,22 +119,37 @@ def test_get_autopayable_invoices_returns_nothing(self):
     def test_pay_autopayable_invoices(self):
         original_outbox_length = len(mail.outbox)
 
-        autopayable_invoice = Invoice.objects.filter(subscription=self.subscription)
-        date_due = autopayable_invoice.first().date_due
+        autopayable_invoice = Invoice.objects.filter(subscription=self.subscription).first()
+        date_due = autopayable_invoice.date_due
+
+        # Keys for idempotent requests can only be used with the same parameters they were first used with.
+        # So introduce randomness in idempotency key to avoid clashes
```

Comment:
Does use of randomness mean there is still some chance of collision if the RNG picks the same numbers twice in a row, or in this case numbers that happen to sum up to the same total (which seems far more likely)? Would it be possible to work around the random element by using a deterministic algorithm? Possibly by seeding the RNG?

It's not quite clear what this is protecting against. Is it trying to make sure several tests in the same test run will not generate the matching idempotency keys, or are is this for multiple test runs within a given day, or something else?

I assume the reason this is needed is because these tests are hitting the Stripe test API. I'm a little nervous about that because our tests could start to fail if their test API is unreliable or if we violate their terms of service and they rate limit us or something.

Reply:
I ran the test locally to determine that there are 3 line items on the invoice. Collisions appear to be fairly common in a quick simulation. Tell me if I got something wrong.
```py
from collections import Counter
from random import randint

sums = [sum(randint(1, 50) for x in range(3)) for y in range(100)]
counter = Counter(sums)
print("# collisions:", sum(n for _, n in counter.most_common() if n > 1))
print("# most common:", counter.most_common(5))

# a typical run on my computer prints
# collisions: 62
# most common: [(71, 5), (51, 3), (75, 3), (96, 3), (67, 3)]
```

Reply:
Another idea would be to inject a UUID or other suitably larges source of randomness into the idempotency key for tests. It might be possible to do that with `mock.patch`.

You are correct, seeding the RNG is not a good solution here. That might have been a reasonable solution if we were trying to protect against duplicate idempotency keys within a single test run. It would have the benefit of removing the randomness across test runs, which reduces likelihood of flaky tests.

Reply:
Even with 10000 I'm seeing 1 collision from time to time with the simulation code. I don't know if that will be enough to prevent the tests from being flaky.

Reply:
Hi Daniel, I don't know how to convert an UUID to a (not too large) integer and make sure the integer is still unique. I tried seed the random generator with 1 `random.seed(1)`, and when use `randint(1,10000)` and I got 0 collisions. How do you feel about this?

Reply:
Hi Daniel, instead of introducing randomness in the invoice total price, I patched `invoice.invoice_number` to return UUID in e20d64c and f0bce14.